bin/wasm-node/rust/src/json_rpc_service.rs

```diff
@@ -1232,30 +1387,38 @@ impl JsonRpcService {
             methods::Response::state_subscribeStorage(&subscription).to_json_response(request_id);
 
         let client = self.clone();
+        let source_disconnected = subscriptions.source_disconnected.clone();
 
         // Spawn a separate task for the subscription.
         (self.tasks_executor.lock().await)(Box::pin(async move {
             futures::pin_mut!(storage_updates);
 
             // Send back to the user the confirmation of the registration.
-            client.send_back(&confirmation);
+            client.send_back(&confirmation, source_id);
 
             loop {
                 // Wait for either a new storage update, or for the subscription to be canceled.
                 let next_block = storage_updates.next();
                 futures::pin_mut!(next_block);
                 match future::select(next_block, &mut unsubscribe_rx).await {
                     future::Either::Left((changes, _)) => {
-                        client.send_back(&smoldot::json_rpc::parse::build_subscription_event(
-                            "state_storage",
-                            &subscription,
-                            &serde_json::to_string(&changes).unwrap(),
-                        ));
+                        if source_disconnected.load(atomic::Ordering::Relaxed) {
+                            break;
+                        }
```

Comment:
What if it's immediately set to `false` afterwards and before you reach the line below?